plugins/outputs/d2l_kinesis/gzipKinesisRecordGenerator.go

```diff
@@ -0,0 +1,199 @@
+package d2lkinesis
+
+import (
+	"bytes"
+	"compress/flate"
+	"compress/gzip"
+
+	"github.com/aws/aws-sdk-go/service/kinesis"
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/serializers"
+)
+
+const gzipHeaderSize = 10
+const gzipFlushBlockSize = 5
+const gzipEndBlockSize = 5
+const gzipFooterSize = 8
+
+func createGZipKinesisRecordGenerator(
+	log telegraf.Logger,
+	maxRecordSize int,
+	pkGenerator partitionKeyGenerator,
+	serializer serializers.Serializer,
+) (kinesisRecordGenerator, error) {
+
+	buffer := bytes.NewBuffer([]byte{})
```

Comment:
Since we we're aiming for `maxRecordSize` and we're probably going to get pretty close in practice would it make sense to set the initial capacity of this array/buffer to `maxRecordSize`? It would be an over-estimate (we'd never fill that up) but it would avoid reallocations as we add stuff to it?

Reply:
So we hold onto the buffer for the entire lifetime of the process. So it will grow and double for the first batch, but stay pretty constant after that.

Reply:
![image](https://user-images.githubusercontent.com/12101120/110727820-f43e0380-81e9-11eb-8634-d9acf07a7556.png)

Reply:
Although I guess the reallocations could put it over as well.

Reply:
In practice it was 1MB over, so I'll make that change in a follow up PR.  Good call.

Reply:
Ah cool. I missed that.